instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/UpcomingRouteObjectsTest.kt

```diff
@@ -172,6 +177,304 @@ class UpcomingRouteObjectsTest : BaseCoreNoCleanUpTest() {
+        stayOnPosition(movedAlongTheRoutePosition)
+        val updateAfterMovedAlongTheRoute = mapboxNavigation.routeProgressUpdates()
+            .filter {
+                it.currentRouteGeometryIndex == 61
+            }
+            .first()
+        val distanceDiffAfterMovedAlongTheRoute = updateAfterMovedAlongTheRoute.distanceTraveled
+        val expectedRoadObjectsAfterMovedAlongTheRoute = originalRoadObjects.drop(1)
+            .map { it.roadObject::class.java to it.distanceToStart!! - distanceDiffAfterMovedAlongTheRoute }
+
+        checkRoadObjects(
+            expectedRoadObjectsAfterMovedAlongTheRoute,
+            updateAfterMovedAlongTheRoute.upcomingRoadObjects
+        )
+
+        mapboxNavigation.routeRefreshController.requestImmediateRouteRefresh()
+
+        mapboxNavigation.routesUpdates()
+            .filter { it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH }
+            .first()
+
+        // distance traveled ~ 22073, second and third road objects are passed
+        val positionAfterFirstRefresh =
+            routes.first().directionsRoute.completeGeometryToPoints()[250]
+        stayOnPosition(positionAfterFirstRefresh)
+        val updateAfterRefresh = mapboxNavigation.routeProgressUpdates()
+            .filter { it.currentRouteGeometryIndex == 249 }
+            .first()
+        val distanceDiffAfterFirstRefresh = updateAfterRefresh.distanceTraveled
+        // refresh geometry_index_start = 1810, so resulting geometry_index_start = 1871
+        // (refresh was made with current_route_geometry_index = 61),
+        // this corresponds to distanceTravelled = 134617.89 ~ 134612.883
+        // (1 geometry index != 1 meter, so 134617.89 is an approximate value),
+        val newIncidentDistanceToStart = 134612.883
+        val newIncidentIndex = 22
+
+        val expectedObjectsAfterFirstRefresh = originalRoadObjects
+            .map {
+                it.roadObject::class.java to it.distanceToStart!! - distanceDiffAfterFirstRefresh
+            }.toMutableList().apply {
+                add(newIncidentIndex, Incident::class.java to newIncidentDistanceToStart - distanceDiffAfterFirstRefresh)
+            }.drop(3)
+
+        checkRoadObjects(expectedObjectsAfterFirstRefresh, updateAfterRefresh.upcomingRoadObjects)
+
+        mockWebServerRule.requestHandlers.removeLast()
+        mockWebServerRule.requestHandlers.add(
+            FailByRequestMockRequestHandler(
+                MockDirectionsRefreshHandler(
+                    "route_with_road_objects_europe",
+                    readRawFileText(context, R.raw.route_with_road_objects_europe_refresh2),
+                    acceptedGeometryIndex = 249
+                )
+            )
+        )
+
+        mapboxNavigation.routeRefreshController.requestImmediateRouteRefresh()
+        mapboxNavigation.routesUpdates()
+            .filter { it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH }
+            .take(2)
```

Comment:
why 2? You requested only 1 refresh, didn't you?

Reply:
oh, it's probably because previous update was caused by refresh and you skip initial value

Reply:
If you extract it to util function, you can check before route request if current value has reason refresh too and then it would be clear why sometimes we wait for 1 and sometimes for two. 
What do you think about it?

Reply:
> oh, it's probably because previous update was caused by refresh and you skip initial value

Yes. And that's always the case here. No other reasons are expected.

> If you extract route refresh to util function, you can check before route request if current routes has reason refresh too and then decide if you need to wait for 1 or 2 updates. I think that can bring more clarity.
What do you think about it?

How do you see this function? I'm not sure we can make it reusable. In lots of places we wait for refresh which happens after timeout, not on demand. So we don't have a "before request" moment.